New features for Cassy GUI

yume-ind opened the pull request:
This pull request contains many changes to the Cassy GUI, including the following:
- Improved formatting of components by adding more margins and padding
- Added a screen for viewing backups
- Added a screen for viewing restores
- Added modals for the user to create and restore backups
- Refactored some code

Vincent left the following comments for consideration in future updates:

General comments :

    Validating the form input will improve user experience. Generally, every-time the user is asked to input something, do not assume he is entering data properly. For example, I was able to enter "foo" as the IP address which should not be permitted.
    If a request fail, it would be nice to be notified about it somehow. For example, if I try to register the cluster but the request fail, the user should know about it.
    Instead of showing a table which does not contain data, you could display a message instead. For example for the "list clusters" view, you can replace the table by the message "There is no registered cluster".

List clusters view:

    Not all the cluster information are displayed in the GUI like the target_ips, keyspaces etc. I don't know if all of these informations should be displayed on the UI but some of it could be useful like the target_ips. If each data element is display, the UI may be cluttered though. One possibility is to let the user parametrize which column he wants to display/hide on the "list cluster" table. Maybe we can check with Yamada-san about this.

Backup view :

    Maybe we could hide or deactivate the "Restore" button if the backup status is not "completed".

feeblefakie commented:
@yume-ind Thank you !

I noticed several things as below.
I think the most critical one is ```As list of backups should be grouped by snapshot_id```. The other ones are relatively minor.

- Overall, time is wrong
- Backups page should show cluster information in details.
- Clicking "View" in Backups page is directed to Restores page
	- Should show details of the backup
- "Backup Type" is not needed. Select "Snapshot" by default
- "Restore Type" is not needed. Select "Node" by default
- "Specifiy nodes" should show a list of IPs registered in the current cluster
- As list of backups should be grouped by snapshot_id
	- incremental backups and cluster backups belong to a snapshot

I think it's better to test it in a multi-node environment.

yume-ind commented:
@feeblefakie 

Thanks for the feedback. I would like to ask for a little bit of clarification:

> Backups page should show cluster information in details.
Do you mean to add a column `Cluster ID` in the Backups list, or to show cluster information in the `Details` page? 

>"Specifiy nodes" should show a list of IPs registered in the current cluster
Does this still need the option to input a new IP, or only to select them from a list?

> As list of backups should be grouped by snapshot_id 
Do you mean I should only show the most recent increment of each snapshot, and show the rest in the `Details` page?

yume-ind commented:
@feeblefakie I pushed the changes to the gui. Please let me know your thoughts so far. I hope I didn't over-complicate it.

yume-ind commented:
@feeblefakie With this recent push, I think I have mostly resolved some of the issues in your comment above.

### Main improvements
- The time is now formatted correctly, so the displayed date is accurate.
- Cluster information is displayed at the top of the Backups screen
- I added the `created_at` and `updated_at` fields to the Backups screen, and removed the `View` button as it is not necessary to have a details screen at the moment.
- The "Restore" button is no longer clickable for `FAILED` backups.
    - I suppose it should also be disabled for `STARTED` ?
- "Backup Type" and "Restore Type" now have a default selected value.
- "Specify Nodes" now has a dropdown menu to select from registered IPs.
- Backups are grouped together by their `snapshot_id`.

### Current issues:
- With the removal of the "View" button, there is currently no way to get to the "Restores" screen.
    - We need to decide if you should be redirected to the Restores screen after restoring from a backup
    - Perhaps we need another way to get to this screen as well, like from the Nav bar.
- The Backups screen shows **all** backups, including all Incremental backups, but it should only show the most recent Incremental backup.

feeblefakie commented on gui/README.md:
````diff
@@ -10,7 +10,7 @@ See the [Getting Started](../docs/getting-started.md) to set up your Cassy confi
 Once you have started Cassy, go to the root folder and run the following command to start the HTTP/1.1 server.
 ```
-./entrypoint -dev
+./entrypoint -mode=dev
````
I think `prod` should be used for normal users.
There might need a explanation for both dev and prod.

feeblefakie commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,130 @@
+        <h4 class="pt-3">Backups in Cluster: {{cluster.cluster_name}}</h4>
+        <ClusterSummary :cluster="cluster"/>
+        <div class="row justify-content-end mb-3 pr-3">
+            <button class="btn btn-primary mx-1" data-toggle="modal" data-target="#registerBackup">Create Backup
```
I'm not sure, but should it be like this ?
```suggestion
            <button class="btn btn-primary mx-1" data-toggle="modal" data-target="#registerBackup">Create a Backup
```
or `Create Backups` ?

yume-ind replied:
It can be as you suggested if you prefer. I think usually buttons like this can omit the "a".

feeblefakie replied:
OK ! Then no `a`. (keeping it as it is is fine)

feeblefakie commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,130 @@
+<template>
+    <div>
+        <h4 class="pt-3">Backups in Cluster: {{cluster.cluster_name}}</h4>
```
```suggestion
        <h4 class="pt-3">{{cluster.cluster_name}}</h4>
```

Very small thing but I think we need a little space under this title. Seems a bit packed.

feeblefakie commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,130 @@
+            >View Restores
+            </button>
+        </div>
+        <table class="table text-center table-bordered">
```
Can you make all the table header color from green to lightgray ?

feeblefakie commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,130 @@
+                            data-toggle="modal"
+                            data-target="#restoreCluster"
+                            @click="$emit('emitRestoreParams', {snapshot_id: e.snapshot_id, backup_type: e.backup_type})"
+                    >Restore
```
I tried a restore, but the `Target IP` of the small window says `all nodes`. I think it should restore it to the IP it took backup from. (so `127.0.0.1` in this case)

![Screen Shot 2020-06-30 at 9 17 35](https://user-images.githubusercontent.com/1052473/86068572-adbcd480-bab2-11ea-8ae1-038d144e6103.png)

yume-ind replied:
I think functionality-wise, the restore behaves as intended (restores to 127.0.0.1). The return values for `Target IP(s)` in the confirmation screen is based on whether or not the request body for restore contains a `target_ips` argument or not. If no, it restores to all nodes that pertain to the backup as in the command:
```
grpcurl -plaintext -d '{"snapshot_id": "SNAPSHOT-ID", "restore_type": 2}' 192.168.0.254:20051 rpc.Cassy.RestoreBackup
```

feeblefakie replied:
Then I think it is weird to restore to all the nodes if you click one backup of a node.
For example, clicking `Restore` of the backup of "192.168.0.2" in a three-node C* cluster (.2, .3, .4), will it restore to all the nodes ?

yume-ind replied:
If you don't select a `target_ip` in the Create Backup screen, it means that it will send a request to the Cassy backend with no argument for `target_ip`. So in a one-node cluster like the one in the screenshot, it will restore to "all nodes", but the only node to restore to is `127.0.0.1`.

In a three-node cluster, if you click `Restore` without specifying a `target_ip`, it will restore to all three nodes. To restore only to `192.168.0.2`, you would need to specify that on the previous screen.

feeblefakie replied:
I understand, but in the current GUI, you will specify a snapshot of a particular target IP by clicking `Restore` of the target IP so I think it should always restore the specified node.  At least for the current GUI, I don't think we need an option to restore to all nodes since it is hard to specify it in the current UI.

yume-ind replied:
Ok. So for now, I will change it to automatically set the target_ip for each entry.

feeblefakie commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,130 @@
+<template>
+    <div>
```
![Screen Shot 2020-06-30 at 9 23 29](https://user-images.githubusercontent.com/1052473/86068884-5ff49c00-bab3-11ea-83e0-56bb280ba8c7.png)
It is now left aligned when you make a browser window bigger, but I think this should be centered.

feeblefakie commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,130 @@
+    <div>
+        <h4 class="pt-3">Backups in Cluster: {{cluster.cluster_name}}</h4>
+        <ClusterSummary :cluster="cluster"/>
+        <div class="row justify-content-end mb-3 pr-3">
```
I think we need a bit more space in between the table and the buttons.
(the table for cluster information and the buttons for `Create Backup` and `View Restore`.)
Ditto for the restore page as well.

yume-ind replied:
Is it okay for the `Clusters` page?

feeblefakie commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,130 @@
+        }
+      },
+      isNotCompleted(entry) {
+        if (entry.backup_type === 1) {
```
We can make it less nested for the for loop by doing the following.

```
if (entry.backup_type !== 1) {
  return entry.status !== 'COMPLETED';
} 

for (let i = 0; i < this.backups_by_snapshot.length; i++) {
  ...
}
```

feeblefakie commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,130 @@
+          for (let i = 0; i < this.backups_by_snapshot.length; i++) {
+            let snapshot = this.backups_by_snapshot[i];
+            for (let j = 0; j < snapshot.length; j++) {
+              if (snapshot[j].snapshot_id === entry.snapshot_id) {
```
This double `if` can be combined ?

feeblefakie commented on gui/src/views/ViewBackups.vue:
```diff
@@ -0,0 +1,117 @@
+          });
+          this.$api.get(`clusters/${this.cluster_id}/backups`)
+          .then((response) => {
+            if (response.status === 200) {
```
If it doesn't return `200`, what will happen ?

This if block can be less nested by doing something like this ?
```
 if (response.status !== 200) {
  // do something ? and return 
}

// do the same things but less nested
this.backups = response.data;

yume-ind replied:
In this case I guess nothing will happen if the request doesn't succeed. I can add some kind of console log or add some feedback on the screen indicating the gui failed to get information from cassy.

yume-ind replied:
Actually, I think if we want to make it do something when the response isn't `200`, it would be a `catch` block like I mentioned in my other comment.

feeblefakie commented on gui/src/components/RegisterCluster.vue:
```diff
@@ -0,0 +1,89 @@
+        this.$api.post('/clusters', {
+          cassandra_host: document.getElementById('cassandraHostInput').value,
+        }).then((response) => {
+          if (response.status === 200) {
```
If it doesn't return `200`, it doesn't happen anything ?
I think some error handling is needed.
The same for the other parts.

yume-ind replied:
I found that the way to catch errors like `400` or `500` etc, you need to add a `catch` block. So in this case and some of the other code, the `.catch(error)` handles cases when the status is not 200.

yume-ind replied:
However, there are some parts in the code where more robust error handling could be good. RegisterCluster currently catches the most common error responses, but could probably use a general `else { ... }` to catch unexpected issues.

feeblefakie replied:
Not fully sure what you said but are you saying error handling should be generalized as a function and it should be used for other similar code ?
If so, I agree.
Anyways, I think we should do error handling; at least showing some error message to somewhere like at the bottom.

Error handling can be done in another PR .

yume-ind replied:
What I mean is this: The axios API seems to interpret codes such as `400` and `500` as errors. So, if you write a line such as `else if (response.status === 500)`, that code will never be reached and it will just throw an error to the console because the request failed.

Instead, you need to `catch` the `400` / `500` status codes like in the code below. This is the way that I've implemented error handling in all the classes so far.

```java
        this.$api.post('/clusters', {
          cassandra_host: document.getElementById('cassandraHostInput').value,
        }).then((response) => {
          if (response.status === 200) {
            this.failed = false;
            $('#registerCluster').modal('hide');
            this.$emit('updateClusterList');
          }
        })
        .catch(error => {                                    // here is where the error handling is done
          this.failed = true;
          if (error.response.data.code === 13) {
            this.error = "Failed to register cluster. Please make sure Cassandra is started."
          } else if (error.response.status === 503) {
            this.error = "Failed to register cluster. Cassy service unavailable."
          }
        })
        ;
```

feeblefakie replied:
Oh, OK. So error handling is done but errors are only shown in console logs.
I think it will be more user friendly if we have some error message space at around the bottom. 
(or at the top. I'm not sure where to put honestly, but there must be some standard in this kind of thing)

@Torch3333  What do you think ?

yume-ind replied:
Not quite. What I mean is, Axios has a default way of handling errors (it sends console logs). However, I am catching that error to tell it what to do. In the example above, I'm setting a data variable called `error` to a message pertaining to what caused the error.

This message is displayed at the bottom of the RegisterCluster modal, like this:
<img width="496" alt="Screen Shot 2020-07-03 at 10 11 14" src="https://user-images.githubusercontent.com/46988771/86421623-afd29d80-bd15-11ea-9d9a-5ec19d963852.png">

Torch3333 commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,132 @@
+            </button>
+        </div>
+        <table class="table text-center table-bordered">
+            <thead class="table-secondary">
```
If there are no backups, the backup table is empty. 
It's personal taste but I find it a bit settling to see an empty table. Without visual cues it makes me wonder if the content was properly loaded or not.
Maybe a text displayed below the header that says something like
"There are no backup yet" or "No backup" would be welcome.
![Screen Shot 2020-07-01 at 18 07 14](https://user-images.githubusercontent.com/3835021/86226111-252f5880-bbc6-11ea-8286-77c7bf7ab5f4.png)

feeblefakie replied:
That's a good idea.

Torch3333 commented on gui/src/components/Restores.vue:
```diff
@@ -0,0 +1,84 @@
+            >View Backups
+            </button>
+        </div>
+        <table class="table table-bordered text-center">
```
Same as the backup table, if there are no backups, should there be a text or a visual cue displayed below the header that says something like
"There are no restores yet"

Torch3333 commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,132 @@
+            </button>
+        </div>
+        <table class="table text-center table-bordered">
+            <thead class="table-secondary">
```
At first glance, it's difficult to distinguish that there are 2 tables displayed in the page. I am wondering if the cluster info table header color should be set to a different color from the backup table.

Torch3333 replied:
The screen I am talking about.
![Screen Shot 2020-07-01 at 18 07 14](https://user-images.githubusercontent.com/3835021/86230098-958ca880-bbcb-11ea-838b-7323908ac3e0.png)

feeblefakie replied:
Good idea. Differentiating by color or structure. 
Having no lines and colors like this might be good. 
![tokens](https://user-images.githubusercontent.com/1052473/86303414-b514e700-bc46-11ea-850a-ed3d99736efa.png)

Or having different structure like this ? (not sure how to remove the header in Markdown)

|    |    |
|-----|--------|
|Cluser ID| test_cluster-xxx    |
|Target IPs  | ["127.0.0.1"]    |
|Keyspaces | ["x", "y", "z"]  |

Torch3333 replied:
Nice, I like the screenshot layout. 
By not putting the cluster info (cluster id, target ips, keyspaces,etc) inside a table, it will put less emphasis on it. The result is that the backup table should pop out.

yume-ind replied:
In terms of keeping the table horizontal, here are a few possible layouts. Do you like any of the following or should I try something else?

v1: Small table
![Screen Shot 2020-07-02 at 15 38 51](https://user-images.githubusercontent.com/46988771/86325037-64b77c80-bc7a-11ea-8558-1b07977a29be.png)
v2: Borderless table
![Screen Shot 2020-07-02 at 15 39 13](https://user-images.githubusercontent.com/46988771/86325058-6b45f400-bc7a-11ea-83f8-dc26bddc7bd3.png)
v3: Normal table, no color
![Screen Shot 2020-07-02 at 15 39 54](https://user-images.githubusercontent.com/46988771/86325066-6da84e00-bc7a-11ea-84c9-750359e2ae11.png)

Torch3333 replied:
I vote for v2.

feeblefakie commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,132 @@
+            >View Restores
+            </button>
+        </div>
+        <table class="table text-center table-bordered">
```
I realized that all table entries might be better to be left aligned instead of centered.
![segment-view](https://user-images.githubusercontent.com/1052473/86303680-8b0ff480-bc47-11ea-9b38-5cdb27aace4c.png)

yume-ind commented:
@feeblefakie I pushed an update to the GUI that does the following:

- Users are no longer able to specify `target_ip` for restores, and it instead is automatically set to the `target_ip` of the row where the button was clicked.
- Added a Toast component for error reporting. At the moment it handles backend errors, like when the GUI is unable to retrieve information from Cassy or something. When such an error occurs you will get a popup notification on the screen to notify you. We can expand on this if needed.
- Tables will now display a "No data found" type entry below them if there is no data to display.

For now, I think that resolves all the comments on this PR.

feeblefakie commented on gui/src/components/CreateBackup.vue:
```diff
@@ -0,0 +1,110 @@
+                            <label for="backupTypeSelect">Select a backup type</label>
+                            <select class="custom-select" id="backupTypeSelect" v-model="backup_type">
+                                <option selected :value="2">Snapshot</option>
+                                <option :value="1">Cluster-Wide</option>
```
```suggestion
                                <option :value="1">Cluster-wide</option>
```
for consistency with [here](https://github.com/scalar-labs/cassy/pull/46/files#diff-1d6c093de68c381c8d652e660be617eaR101).

feeblefakie commented on gui/src/components/Backups.vue:
```diff
@@ -0,0 +1,140 @@
+<template>
+    <div>
+        <h4 class="py-4">Backups in Cluster: {{cluster.cluster_name}}</h4>
```
```suggestion
        <h4 class="py-4">Backups</h4>
```

feeblefakie commented on gui/src/components/ClusterSummary.vue:
```diff
@@ -0,0 +1,29 @@
+    <table class="table table-borderless text-left">
+        <thead>
+        <tr>
+            <th scope="col">Cluster ID</th>
```
```suggestion
            <th scope="col">Cluster ID</th>
            <th scope="col">Cluster Name</th>
```
I just realized that cluster name should be here instead of in the title.

feeblefakie commented on gui/src/components/Restores.vue:
```diff
@@ -0,0 +1,92 @@
+<template>
+    <div>
+        <h4 class="py-4">Restore Statuses in Cluster: {{cluster.cluster_name}}</h4>
```
```suggestion
        <h4 class="py-4">Restore Statuses</h4>
```

feeblefakie commented on gui/src/App.vue:
```diff
@@ -1,24 +1,94 @@
-        </nav>
-        <router-view/>
+        </div>
         <footer class="fixed-bottom">
```
I think we can delete the footer.

yume-ind commented:
@feeblefakie Regarding the `./entrypoint -mode=dev` issue - for the time being, that is the way to start the GUI. I think in the future the GUI will be run differently, so that `./entrypoint` can run normally on port 8080. I can try to include that in this PR if you want. But basically `./entrypoint -mode=dev` is not for basic users, but it's just a temporary thing so that we can use `npm run serve` and benefit from hot reload during development without conflicting ports.
